app/scripts/lib/Stores/ExtensionStore.ts

```diff
@@ -0,0 +1,196 @@
+ *
+ * @param obj - The object to check
+ * @returns
+ */
+function isEmpty(
+  obj: MetaMaskStorageStructure | EmptyState,
+): obj is EmptyState {
+  return Object.keys(obj).length === 0;
+}
+
+/**
+ * An implementation of the MetaMask Extension BaseStore system that uses the
+ * browser.storage.local API to persist and retrieve state.
+ */
+export class ExtensionStore extends BaseStore {
+  isSupported: boolean;
+
+  stateCorruptionDetected: boolean;
+
+  dataPersistenceFailing: boolean;
+
+  isExtensionInitialized: boolean;
+
+  mostRecentRetrievedState: MetaMaskStorageStructure | null;
+
+  migrator: Migrator;
+
+  constructor({ migrator }: { migrator: Migrator }) {
+    super();
+    this.stateCorruptionDetected = false;
+    this.isSupported = Boolean(browser.storage.local);
+    if (!this.isSupported) {
+      log.error('Storage local API not available.');
+    }
+    // we use this flag to avoid flooding sentry with a ton of errors:
+    // once data persistence fails once and it flips true we don't send further
+    // data persistence errors to sentry
+    this.dataPersistenceFailing = false;
+    this.mostRecentRetrievedState = null;
+    this.isExtensionInitialized = false;
+    this.migrator = migrator;
+  }
+
+  /**
+   * Persists the MetaMask state tree on the 'data' key of the
+   * browser.storage.local API. This function will first do some sanity checks
+   * to determine if it is likely for the operation to succeed. It will throw
+   * an error in the following cases.
+   * 1. The browser does not support the browser.storage.local API.
+   * 2. No state object was provided to the function. As more of the codebase
+   * is migrated to TypeScript, this should become less of a possibility but
+   * never impossible.
+   * 3. The metadata property is not set on the class which is required before
+   * setting state. This ensures the 'meta' key of the storage is set.
+   *
+   * @param state - The state to persist to the data key of the local store
+   * @returns void
+   */
+  async set(state: IntermediaryStateType) {
+    if (!this.isSupported) {
+      throw new Error(
+        'Metamask- cannot persist state to local store as this browser does not support this action',
+      );
+    }
+    if (!state) {
+      throw new Error('MetaMask - updated state is missing');
+    }
+    if (!this.metadata) {
+      throw new Error(
+        'MetaMask - metadata must be set on instance of ExtensionStore before calling "set"',
+      );
+    }
+    try {
+      // we format the data for storage as an object with the "data" key for the controller state object
+      // and the "meta" key for a metadata object containing a version number that tracks how the data shape
+      // has changed using migrations to adapt to backwards incompatible changes
+      await this.#set({ data: state, meta: this.metadata });
+      if (this.dataPersistenceFailing) {
+        this.dataPersistenceFailing = false;
```

Comment:
Looking at this makes me think we're mixing concerns here. This logic is helpful for our "primary store" no matter what storage backend we're using. But it's implemented in a storage-backend-specific class.

Perhaps instead what we want is a single "PersistedStore" class, and then separate "BaseStorage", "ExtensionStorage", and "NetworkStorage" classes that are used by the "PersistedStore". The "PersistedStore" would handle metadata/version management, fallback logic, tracking "failure" state, etc, while the storage classes would _just_ get/set, and could be easily switched around without any other logic changes.